cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java

```diff
@@ -136,6 +136,20 @@ public void run() {
     return result;
   }
 
+  protected void complete() {
+    if (!state.compareAndSet(State.ALIVE, State.COMPLETED)) {
+      LOG.debug("Cannot transit to COMPLETED state from {} state: {} {}", state.get(), programName, runId);
```

Comment:
Hum.. maybe there could be race between stopping and completed (someone call controller.stop() at about the same time that the program finished). Probably ok to log it as debug.